File: app/src/main/java/com/mbestavros/geometricweather/basic/model/weather/Temperature.java
package com.mbestavros.geometricweather.basic.model.weather;

import android.content.Context;

import androidx.annotation.Nullable;

import java.io.Serializable;

import com.mbestavros.geometricweather.basic.model.option.unit.TemperatureUnit;
import com.mbestavros.geometricweather.settings.SettingsOptionManager;

/**
 * Temperature.
 * default unit : {@link com.mbestavros.geometricweather.basic.model.option.unit.TemperatureUnit#C}
 * */
public class Temperature implements Serializable {

    private int temperature;
    @Nullable private Integer realFeelTemperature;
    @Nullable private Integer realFeelShaderTemperature;
    @Nullable private Integer apparentTemperature;
    @Nullable private Integer windChillTemperature;
    @Nullable private Integer wetBulbTemperature;
    @Nullable private Integer degreeDayTemperature;

    public Temperature(int temperature,
                       @Nullable Integer realFeelTemperature,
                       @Nullable Integer realFeelShaderTemperature,
                       @Nullable Integer apparentTemperature,
                       @Nullable Integer windChillTemperature,
                       @Nullable Integer wetBulbTemperature,
                       @Nullable Integer degreeDayTemperature) {
        this.temperature = temperature;
        this.realFeelTemperature = realFeelTemperature;
        this.realFeelShaderTemperature = realFeelShaderTemperature;
        this.apparentTemperature = apparentTemperature;
        this.windChillTemperature = windChillTemperature;
        this.wetBulbTemperature = wetBulbTemperature;
        this.degreeDayTemperature = degreeDayTemperature;
    }

    public int getTemperature() {
        return temperature;
    }

    @Nullable
    public Integer getRealFeelTemperature() {
        return realFeelTemperature;
    }

    @Nullable
    public Integer getRealFeelShaderTemperature() {
        return realFeelShaderTemperature;
    }

    @Nullable
    public Integer getApparentTemperature() {
        return apparentTemperature;
    }

    @Nullable
    public Integer getWindChillTemperature() {
        return windChillTemperature;
    }

    @Nullable
    public Integer getWetBulbTemperature() {
        return wetBulbTemperature;
    }

    @Nullable
    public Integer getDegreeDayTemperature() {
        return degreeDayTemperature;
    }

    @Nullable
    public String getTemperature(Context context, TemperatureUnit unit) {
        return getTemperature(context, temperature, unit);
    }

    @Nullable
    public String getShortTemperature(Context context, TemperatureUnit unit) {
        return getShortTemperature(context, temperature, unit);
    }

    @Nullable
    public String getRealFeelTemperature(Context context, TemperatureUnit unit) {
        return getTemperature(context, realFeelTemperature, unit);
    }

    @Nullable
    public String getShortRealFeeTemperature(Context context, TemperatureUnit unit) {
        return getShortTemperature(context, realFeelTemperature, unit);
    }

    @Nullable
    public static String getTemperature(Context context, @Nullable Integer temperature, TemperatureUnit unit) {
        if (temperature == null) {
            return null;
        }
        return unit.getTemperatureText(context, temperature);
    }

    @Nullable
    public static String getShortTemperature(Context context,
                                             @Nullable Integer temperature, TemperatureUnit unit) {
        if (temperature == null) {
            return null;
        }
        return unit.getShortTemperatureText(context, temperature);
    }

    @Nullable
    public static String getTrendTemperature(Context context,
                                             @Nullable Integer night, @Nullable Integer day,
                                             TemperatureUnit unit) {
        if (night == null || day == null) {
            return null;
        }
        if (SettingsOptionManager.getInstance(context).isExchangeDayNightTempEnabled()) {
            return getShortTemperature(context, day, unit) + "/" + getShortTemperature(context, night, unit);
        } else {
            return getShortTemperature(context, night, unit) + "/" + getShortTemperature(context, day, unit);
        }
    }

    public boolean isValid() {
        return realFeelTemperature != null
                || realFeelShaderTemperature != null
                || apparentTemperature != null
                || windChillTemperature != null
                || wetBulbTemperature != null
                || degreeDayTemperature != null;
    }
}

File: app/src/main/java/org/breezyweather/common/basic/models/weather/Temperature.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.common.basic.models.weather

import android.content.Context
import org.breezyweather.common.basic.models.options.unit.TemperatureUnit
import java.io.Serializable

/**
 * Temperature.
 * default unit : [TemperatureUnit.C]
 */
data class Temperature(
    val temperature: Float? = null,
    val realFeelTemperature: Float? = null,
    val realFeelShaderTemperature: Float? = null,
    val apparentTemperature: Float? = null,
    val windChillTemperature: Float? = null,
    val wetBulbTemperature: Float? = null
) : Serializable {

    companion object {
        fun getTemperature(context: Context, temperature: Float?, unit: TemperatureUnit, decimals: Int = 1): String? {
            return temperature?.let { unit.getValueText(context, it, decimals) }
        }

        fun getShortTemperature(context: Context, temperature: Float?, unit: TemperatureUnit): String? {
            return temperature?.let { unit.getShortValueText(context, it) }
        }

        fun getTrendTemperature(context: Context, night: Float?, day: Float?, unit: TemperatureUnit): String? {
            if (night == null || day == null) {
                return null
            }
            return getShortTemperature(context, day, unit) + "/" + getShortTemperature(context, night, unit)
        }
    }

    val feelsLikeTemperature: Float? = realFeelTemperature ?: realFeelShaderTemperature
        ?: apparentTemperature ?: windChillTemperature ?: wetBulbTemperature

    fun getTemperature(context: Context, unit: TemperatureUnit, decimals: Int = 1): String? {
        return getTemperature(context, temperature, unit, decimals)
    }

    fun getShortTemperature(context: Context, unit: TemperatureUnit): String? {
        return getShortTemperature(context, temperature, unit)
    }

    fun getFeelsLikeTemperature(context: Context, unit: TemperatureUnit, decimals: Int = 1): String? {
        return getTemperature(context, feelsLikeTemperature, unit, decimals)
    }

    fun getShortFeelsLikeTemperature(context: Context, unit: TemperatureUnit): String? {
        return getShortTemperature(context, feelsLikeTemperature, unit)
    }
}
